layers/layers.go

```diff
@@ -0,0 +1,273 @@
+			if err := ls.driver.Remove(layer.cacheID); err != nil {
+				logrus.Errorf("Error cleaning up cache layer %s: %v", layer.cacheID, err)
+			}
+		}
+	}()
+
+	digester := digest.Canonical.New()
+	tr := io.TeeReader(ts, digester.Hash())
+
+	layer.size, err = ls.driver.ApplyDiff(layer.cacheID, pid, archive.Reader(tr))
+	if err != nil {
+		return nil, err
+	}
+
+	layer.tarStreamer = func() (io.Reader, error) {
+		archiver, err := ls.driver.Diff(layer.cacheID, pid)
+		return io.Reader(archiver), err
+	}
+
+	var layerHash digest.Digest
+	if layer.parent == nil {
+		layerHash = digester.Digest()
+	} else {
+		layerHashParts := digester.Digest().String() + " " + digest.Digest(layer.parent.address).String()
+		layerHash, err = digest.FromBytes([]byte(layerHashParts))
+		if err != nil {
+			return nil, err
+		}
+	}
+	layer.address = LayerAddress(layerHash)
+
+	ls.layerL.Lock()
+	defer ls.layerL.Unlock()
+
+	if existingLayer, ok := ls.layerMap[layer.address]; ok {
+		// Set error for cleanup, but do not return
+		err = errors.New("layer already exists")
+		return existingLayer, nil
+	}
+
+	ls.layerMap[layer.address] = layer
+
+	// TODO: Persist mapping update to disk
+
+	return layer, nil
+}
+
+func (ls *layerStore) Get(l LayerAddress) (Layer, error) {
+	ls.layerL.Lock()
+	defer ls.layerL.Unlock()
+
+	layer, ok := ls.layerMap[l]
+	if !ok {
+		return nil, ErrLayerDoesNotExist
+	}
+	// TODO: Add name here to keep reference?
+	return layer, nil
+}
+
+func (ls *layerStore) Put(l LayerAddress) error {
+	// TODO: Release reference, attempt garbage collections
+	// NOTE: If put is called on layer before layers have all
+	// been fully retrieved via Get, layers may unintentionally
+	// be removed.
+	return nil
+}
+
+func (ls *layerStore) Mount(id string, parent LayerAddress) (RWLayer, error) {
+	ls.mountL.Lock()
+	defer ls.mountL.Unlock()
+	if m, ok := ls.mounts[id]; ok {
+		return m, nil
+	}
+
+	ls.layerL.Lock()
+	defer ls.layerL.Unlock()
+
+	var pid string
+	var p *cacheLayer
+	if string(parent) != "" {
+		l, ok := ls.layerMap[parent]
+		if !ok {
+			return nil, ErrLayerDoesNotExist
+		}
+		p = l
+		pid = l.cacheID
+	}
+
+	mount := &mountedLayer{
+		parent:  p,
+		mountID: stringid.GenerateRandomID(),
+	}
+
+	if err := ls.driver.Create(mount.mountID, pid); err != nil {
+		return nil, err
+	}
+
+	mount.tarStreamer = func() (io.Reader, error) {
+		archiver, err := ls.driver.Diff(mount.mountID, pid)
+		return io.Reader(archiver), err
+	}
+
+	dir, err := ls.driver.Get(mount.mountID, "")
+	if err != nil {
+		// TODO: Cleaup
+		return nil, err
+	}
+	mount.path = dir
+
+	ls.mounts[id] = mount
+
+	// TODO: Persist mapping update to disk
+
+	return mount, nil
+}
+
+func (ls *layerStore) Unmount(id string) error {
+	ls.mountL.Lock()
+	defer ls.mountL.Unlock()
+
+	m := ls.mounts[id]
+	if m == nil {
+		return errors.New("mount does not exist")
+	}
+
+	delete(ls.mounts, id)
+
+	return ls.driver.Remove(m.mountID)
```

Comment:
Plain unmount should be `driver.Put`. Do we need a separate way to handle the removing? Btw, I think currently its not safe to call mount/unmount multiple times, Michael is been trying to fix that for aufs https://github.com/docker/docker/pull/15868